### PR TITLE
Qt1 377 view draft apprenticeship add rpl

### PR DESCRIPTION
**Pull request by alireid**

### CODE REVIEW

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web.UnitTests/Mappers/DraftApprenticeshipTests/WhenIMapViewDraftApprenticeshipDetailsToViewModel.cs [hunk 1]
```diff
@@ -94,9 +100,9 @@ public async Task ThenDateOfBirthIsMappedCorrectly()
             Assert.AreEqual(_apiResponse.DateOfBirth, result.DateOfBirth);
         }
 
-        [TestCase(DeliveryModel.Regular)]
-        [TestCase(DeliveryModel.PortableFlexiJob)]
-        public async Task ThenDeliveryModelIsMappedCorrectly(DeliveryModel dm)
+        [TestCase(Infrastructure.OuterApi.Types.DeliveryModel.Regular)]
+        [TestCase(Infrastructure.OuterApi.Types.DeliveryModel.PortableFlexiJob)]
+        public async Task ThenDeliveryModelIsMappedCorrectly(Infrastructure.OuterApi.Types.DeliveryModel dm)
```
Comment (Pauljgraham) [inline src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web.UnitTests/Mappers/DraftApprenticeshipTests/WhenIMapViewDraftApprenticeshipDetailsToViewModel.cs:105]: a possibility is to use this in the using statement area

using DeliveryModel = SFA.DAS.ProviderCommitments.Infrastructure.OuterApis.Types.DeliveryModel;

to reduce the noise, but it's up to you

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments/Infrastructure/OuterApi/OuterApiService.cs [hunk 2]
```diff
@@ -106,5 +106,10 @@ public async Task<GetCohortDetailsResponse> GetCohortDetails(long providerId, lo
         {
             return await _outerApiClient.Get<GetCohortDetailsResponse>(new GetCohortDetailsRequest(providerId, cohortId));
         }
+
+        public async Task<GetDraftApprenticeshipResponse> GetDraftApprenticeship(long providerId, long cohortId, long draftApprenticeshipId)
+        {
+            return await _outerApiClient.Get<GetDraftApprenticeshipResponse>(new GetDraftApprenticeshipRequest(providerId, cohortId, draftApprenticeshipId));
+        }
```
Comment (Pauljgraham) [inline src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments/Infrastructure/OuterApi/OuterApiService.cs:113]: I like putting this call here, but I don't think weu are ever calling it. Your code is calling the Outer API directly in the mapper

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web.UnitTests/Mappers/DraftApprenticeshipTests/WhenIMapViewDraftApprenticeshipDetailsToViewModel.cs [hunk 3]
```diff
@@ -19,27 +23,29 @@ public class WhenIMapViewDraftApprenticeshipDetailsToViewModel
...
-                .ReturnsAsync(_apiResponse);
 
             commitmentsApiClient.Setup(x => x.GetTrainingProgramme(_apiResponse.CourseCode, It.IsAny<CancellationToken>()))
                 .ReturnsAsync(_apiTrainingProgrammeResponse);
 
-            _mapper = new ViewDraftApprenticeshipViewModelMapper(commitmentsApiClient.Object);
+            OuterApiClient = new Mock<IOuterApiClient>();
+            OuterApiClient.Setup(x => x.Get<Infrastructure.OuterApi.Requests.Cohorts.GetDraftApprenticeshipResponse>(It.IsAny<GetDraftApprenticeshipRequest>()))
+                .ReturnsAsync(_apiResponse);
+
```
Comment (Pauljgraham) [inline src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web.UnitTests/Mappers/DraftApprenticeshipTests/WhenIMapViewDraftApprenticeshipDetailsToViewModel.cs:47]: We have updated this file using the IOuterAPIClient but I think it should be using the IApprovalService. Also the new fields are not supported by any tests

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web.UnitTests/Mappers/DraftApprenticeshipTests/WhenIMapViewDraftApprenticeshipDetailsToViewModel.cs [hunk 4]
```diff
@@ -6,8 +6,12 @@
 using NUnit.Framework;
 using SFA.DAS.CommitmentsV2.Api.Client;
 using SFA.DAS.CommitmentsV2.Api.Types.Responses;
+using SFA.DAS.CommitmentsV2.Shared.Models;
 using SFA.DAS.CommitmentsV2.Types;
 using SFA.DAS.CommitmentsV2.Types.Dtos;
+using SFA.DAS.ProviderCommitments.Infrastructure.OuterApi;
+using SFA.DAS.ProviderCommitments.Infrastructure.OuterApi.Requests;
+using SFA.DAS.ProviderCommitments.Infrastructure.OuterApi.Requests.Cohorts;
```
Comment (Pauljgraham) [inline src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web.UnitTests/Mappers/DraftApprenticeshipTests/WhenIMapViewDraftApprenticeshipDetailsToViewModel.cs:14]: Not sure if you get this but I can see unused "using statements", which I can then remove.  There are several of those here (not a big issue)

**Review by Pauljgraham — CHANGES_REQUESTED** [review]
Code looks good, a few issues,  worth addressing

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments/Infrastructure/OuterApi/Requests/Cohorts/GetDraftApprenticeshipRequest.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            ProviderId = providerId;
+            CohortId = cohortId;
+            DraftApprenticeshipId = draftApprenticeshipId;
+        }
+
+        public string GetUrl => $"provider/{ProviderId}/unapproved/{CohortId}/apprentices/{DraftApprenticeshipId}/view";
+    }
+
+    public class GetDraftApprenticeshipResponse
+    {
```
[the remark this answers is not in the document]
Comment (alireid) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments/Infrastructure/OuterApi/Requests/Cohorts/GetDraftApprenticeshipRequest.cs:23]: Good thinking, fixed!

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments/Infrastructure/OuterApi/IOuterApiService.cs [hunk 6]
```diff
@@ -29,5 +29,6 @@ public interface IOuterApiService
         Task<CreatePriorLearningDataResponse> UpdatePriorLearningData(long providerId, long cohortId, long draftApprenticeshipId, CreatePriorLearningDataRequest request);
         Task<GetPriorLearningSummaryQueryResult> GetPriorLearningSummary(long providerId, long cohortId, long apprenticeshipId);
         Task<GetCohortDetailsResponse> GetCohortDetails(long providerId, long cohortId);
+        Task<GetDraftApprenticeshipResponse> GetDraftApprenticeship(long providerId, long cohortId, long draftApprenticeshipId);
```
[the remark this answers is not in the document]
Comment (alireid) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments/Infrastructure/OuterApi/IOuterApiService.cs:32]: Good thinking, fixed!

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Views/DraftApprenticeship/ViewDraftApprenticeship.cshtml [hunk 7]
```diff
@@ -10,151 +10,223 @@
...
+                    }
+                    if (Model.DurationReducedByHours.HasValue)
+                    {
+                        <div class="govuk-summary-list__row">
+                            <dt class="govuk-summary-list__key">
+                                Off-the-job training time reduction due to prior learning
+                            </dt>
+                            <dd class="govuk-summary-list__value">
+                                @Model.DurationReducedByHours.GetValueOrDefault().ToString("N0") @PluraliseString(Model.DurationReducedByHours.GetValueOrDefault(), "hour")
+                            </dd>
```
[the remark this answers is not in the document]
Comment (alireid) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Views/DraftApprenticeship/ViewDraftApprenticeship.cshtml:196]: Cleaner, made that change, thanks

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web.UnitTests/Mappers/DraftApprenticeshipTests/WhenIMapViewDraftApprenticeshipDetailsToViewModel.cs (hunk 1 above)
Comment (alireid) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web.UnitTests/Mappers/DraftApprenticeshipTests/WhenIMapViewDraftApprenticeshipDetailsToViewModel.cs:105]: Initially i did something similar, but as we are also including the commitments nuget types VS has a hard time working out which ones to use. Decided to be safe here with full namespace

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web.UnitTests/Mappers/DraftApprenticeshipTests/WhenIMapViewDraftApprenticeshipDetailsToViewModel.cs (hunk 4 above)
Comment (alireid) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web.UnitTests/Mappers/DraftApprenticeshipTests/WhenIMapViewDraftApprenticeshipDetailsToViewModel.cs:14]: Done a tidy on my side

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments/Infrastructure/OuterApi/Requests/Cohorts/GetDraftApprenticeshipRequest.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
+﻿using SFA.DAS.ProviderCommitments.Infrastructure.OuterApi.Types;
+using System;
+
+namespace SFA.DAS.ProviderCommitments.Infrastructure.OuterApi.Requests.Cohorts
+{
+    public class GetDraftApprenticeshipRequest : IGetApiRequest
+    {
```
[the remark this answers is not in the document]
Comment (alireid) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments/Infrastructure/OuterApi/Requests/Cohorts/GetDraftApprenticeshipRequest.cs:7]: nice, yep fixed

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Views/DraftApprenticeship/ViewDraftApprenticeship.cshtml [hunk 9]
```diff
@@ -10,151 +10,223 @@
...
+                @if (Model.RecognisePriorLearning == true)
+                {
+                    if (Model.TrainingTotalHours.HasValue)
+                    {
+                        <div class="govuk-summary-list__row">
+                            <dt class="govuk-summary-list__key">
+                                Total off-the-job training time for this apprenticeship standard
+                            </dt>
+                            <dd class="govuk-summary-list__value">
+                                £@Model.PriceReducedBy.GetValueOrDefault().ToString("N0")
```
[the remark this answers is not in the document]
Comment (alireid) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Views/DraftApprenticeship/ViewDraftApprenticeship.cshtml]: nice fixed

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Views/DraftApprenticeship/ViewDraftApprenticeship.cshtml [hunk 10]
```diff
@@ -10,151 +10,223 @@
...
+                        </div>
+                    }
+                    if (Model.PriceReducedBy.HasValue)
+                    {
+                        <div class="govuk-summary-list__row">
+                            <dt class="govuk-summary-list__key">
+                                Total price reduction due to RPL
+                            </dt>
+                            <dd class="govuk-summary-list__value">
+                                £@Model.PriceReducedBy.GetValueOrDefault().ToString("N0")
```
[the remark this answers is not in the document]
Comment (alireid) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Views/DraftApprenticeship/ViewDraftApprenticeship.cshtml:218]: fixed

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Mappers/ViewDraftApprenticeshipViewModelMapper.cs [hunk 11]
```diff
@@ -1,25 +1,30 @@
...
             _commitmentsApiClient = commitmentsApiClient;
+            _outerApiClient = outerApiClient;
         }
 
         public async Task<ViewDraftApprenticeshipViewModel> Map(DraftApprenticeshipRequest source)
         {
-            var draftApprenticeship = await _commitmentsApiClient.GetDraftApprenticeship(source.CohortId, source.DraftApprenticeshipId);
+            var draftApprenticeshipTask = _outerApiClient.Get<GetDraftApprenticeshipResponse>(new GetDraftApprenticeshipRequest(source.ProviderId, source.CohortId, source.DraftApprenticeshipId));
+
+            await Task.WhenAll(draftApprenticeshipTask);
```
[the remark this answers is not in the document]
Comment (alireid) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Mappers/ViewDraftApprenticeshipViewModelMapper.cs:25]: Not sure why I did that, fixed and awaited

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments/Infrastructure/OuterApi/OuterApiService.cs (hunk 2 above)
Comment (alireid) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments/Infrastructure/OuterApi/OuterApiService.cs:113]: yep, agreed, fixed

**Review by Pauljgraham — APPROVED** [review]

**Review by Pauljgraham — DISMISSED** [review]

**Review by chrisfoster76 — APPROVED** [review]

**Review by Pauljgraham — DISMISSED** [review]

**Review by chrisfoster76 — DISMISSED** [review]

**Review by chrisfoster76 — DISMISSED** [review]

**Review by chrisfoster76 — APPROVED** [review]

**Review by Pauljgraham — APPROVED** [review]

**Pull request merged**